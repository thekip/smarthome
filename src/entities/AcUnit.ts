import { debounce, pick } from 'lodash';
import { ModbusMaster } from 'modbus-rtu';
import { ModbusCrcError, ModbusResponseTimeout } from 'modbus-rtu/lib/errors';
import { EventEmitter } from '../libs/event-emitter/event-emitter';

export interface AcUnitDto {
  enabled: boolean;
  mode: number;
  fanSpeed: number;
  tempSetpoint: number;
  ambientTemp: number;
}

export enum AcRegister {
  ENABLED = 0,
  UNIT_MODE = 1,
  FAN_SPEED = 2,
  TEMP_SETPOINT = 4,
  AMBIENT_TEMP_AC_UNIT = 5,
  AMBIENT_TEMP_EXTERNAL = 22,
  AC_ACTUAL_SETPOINT_TEMP = 23,
}

export const AMBIENT_TEMP_DEFAULT = -32768;

export enum AcMode {
  AUTO = 0,
  HEAT = 1,
  DRY = 2,
  FAN = 3,
  COOL = 4,
}

export class AcUnit {
  public enabled = false;
  public mode: AcMode = 0;
  public fanSpeed = 0;
  public tempSetpoint = 0;
  public ambientTemp = 0;

  // refer to documentation http://www.intesis.com/pdf/IntesisBox_ME_AC_MBS_1_manual_eng.pdf  p3.2.4
  public $ambientTempAcUnit = 0;
  public $actualSetpoint = 0;

  public onChange = new EventEmitter();
  private triggerChangeEvent = debounce(this.onChange.emit.bind(this.onChange));

  constructor(
    private modbusMaster: ModbusMaster,
    private modbusAddr: number,
  ) {
  }

  public update() {
    return this.modbusMaster.readHoldingRegisters(this.modbusAddr, 0, 24).then((data) => {
      this.enabled = !!data[AcRegister.ENABLED];
      this.mode = data[AcRegister.UNIT_MODE];
      this.fanSpeed = data[AcRegister.FAN_SPEED];
      this.tempSetpoint = data[AcRegister.TEMP_SETPOINT];
      this.ambientTemp = data[AcRegister.AMBIENT_TEMP_EXTERNAL];

      this.$actualSetpoint = data[AcRegister.AC_ACTUAL_SETPOINT_TEMP];
      this.$ambientTempAcUnit = data[AcRegister.AMBIENT_TEMP_AC_UNIT];
    }).catch(ModbusCrcError, ModbusResponseTimeout, () => {
      // do nothing
    });
  }

  public setMode(mode: AcMode) {
    if (this.mode === mode) {
      return;
    }
    this.mode = mode;
    return this.write(AcRegister.UNIT_MODE, mode);
  }

  public setTempSetpoint(setpoint: number) {
    if ((+this.tempSetpoint) === (+setpoint)) {
      return;
    }
    this.tempSetpoint = setpoint;
    return this.write(AcRegister.TEMP_SETPOINT, setpoint);
  }

  public resetControls() {
    this.setDefaultAmbientTemp();
  }

  public setAmbientTemp(temp: number) {
    if (this.ambientTemp === temp) {
      return;
    }
    this.ambientTemp = temp;
    return this.write(AcRegister.AMBIENT_TEMP_EXTERNAL, temp);
  }

  public setDefaultAmbientTemp() {
    return this.setAmbientTemp(AMBIENT_TEMP_DEFAULT);
  }

  public setEnabled(value: boolean) {
    if (this.enabled === value) {
      return;
    }
    this.enabled = value;
    return this.write(AcRegister.ENABLED, +value);
  }

  public toString() {
    return 'Status: ' + (this.enabled ? 'on' : 'off ') +
      '; Ambient temp: ' + this.ambientTemp + 'C; Setpoint: ' + this.tempSetpoint + 'C; \n\r' +
      'Real AC ambient temp: ' + this.$ambientTempAcUnit + 'C; Real AC setpoint: ' + this.$actualSetpoint + 'C;';
  }

  public getDto(): AcUnitDto {
    return pick(this, ['enabled', 'mode', 'fanSpeed', 'tempSetpoint', 'ambientTemp']) as any;
  }

  private write(reg: AcRegister, value: number) {
    this.triggerChangeEvent();
    return this.modbusMaster.writeSingleRegister(this.modbusAddr, reg, value);
  }
}
