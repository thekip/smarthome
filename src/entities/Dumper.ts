import { VavController } from './VavController';

const FULL_CLOSED = 100;
const FULL_OPENED = 0;

export class Dumper {
  private position = 0;

  constructor(
    private port: number,
    private vavCtrl: VavController,
  ) {}

  get isClosed() {
    return this.position >= FULL_CLOSED;
  }

  get isOpened() {
    return this.position === FULL_OPENED;
  }

  public close() {
    this.setPosition(100);
  }

  public open() {
    this.setPosition(0);
  }

  /**
   * Set dumper position in percent. 100% is fully closed
   */
  public setPosition(percent: number) {
    this.position = percent;
    this.vavCtrl.setDumperPosition(this.port, this.position);
  }

  public getPosition(): number {
    return this.position;
  }
}
