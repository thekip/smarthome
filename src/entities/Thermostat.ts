import { ModbusMaster } from 'modbus-rtu';
import { DeviceConnection } from '../libs/device-connection';
import { EventEmitter } from '../libs/event-emitter';
import { Log } from '../libs/log';
import { ModbusCrcError, ModbusResponseTimeout } from 'modbus-rtu/lib/errors';
import { RetriableAction } from '../libs/retriable-action';
import { QueueMap, WriteQueue } from '../libs/write-queue';
import Bluebird = require('bluebird');

export const DEFAULT_SETPOINT = 24;

export const ThermostatRegister = {
  ENABLED: 0,
  FAN_SPEED: 1,
  MODE: 2,
  ROOM_TEMP: 3,
  TEMP_SETPOINT: 4,
  MANUAL_WEEKLY_PROG: 5,

  MINUTE_SECOND: 6,
  WEEK_HOUR: 7,
};

export const ENABLED_VALUE = 165;
export const DISABLED_VALUE = 90;

export class Thermostat {
  public enabled = false;
  public roomTemp: number = -1;
  public tempSetpoint = DEFAULT_SETPOINT;

  /**
   * Emits only if hardware data changed
   */
  public onChange = new EventEmitter();

  private currentData: number[] = [];
  private connection = new DeviceConnection();
  private writeQueues = new QueueMap( () => new WriteQueue());

  constructor(
    private modbusMaster: ModbusMaster,
    private modbusAddr: number,
    noWatch = false,
  ) {

    this.connection.onConnectionLost.subscribe((err) => {
      Log.error('Lost connection to thermostat ' + this.modbusAddr, 'Error:', err.message);
    });

    this.connection.onConnectionRestore.subscribe(() => {
      Log.error('Connection restored. Thermostat ' + this.modbusAddr);
    });

    if (!noWatch) {
      this.watch();
    }
  }

  public update(): Bluebird<number[]> {
    return this.modbusMaster.readHoldingRegisters(this.modbusAddr, 0, 6).then((data) => {
      this.connection.success();

      if (this.writeQueues.length) {
        // discard values from thermostat, if there are tasks
        return this.currentData;
      }

      this.roomTemp = data[ThermostatRegister.ROOM_TEMP] / 2;
      this.enabled = data[ThermostatRegister.ENABLED] === ENABLED_VALUE;
      this.tempSetpoint = data[ThermostatRegister.TEMP_SETPOINT] / 2;

      // check, whether data is changed or not
      if (this.isHwDataChanged(data)) {
        this.onChange.emit();
      }
      this.currentData = data.slice(0); // clone data array

      return data;
    }).catch(ModbusCrcError, ModbusResponseTimeout, (err: Error) => {
      this.connection.error(err);
      return [];
    });
  }

  public toString(): string {
    return `Room temp: ${this.roomTemp} C;`;
  }

  public setEnable(value: boolean): void {
    this.enabled = !!value;
    const modbusValue = !value ? DISABLED_VALUE : ENABLED_VALUE;
    this.currentData[ThermostatRegister.ENABLED] = modbusValue;
    this.writeToThermostat(9, modbusValue);

  }

  public setTempSetpoint(temp: number): void {
    this.tempSetpoint = temp;
    this.currentData[ThermostatRegister.TEMP_SETPOINT] = temp * 2;

    this.writeToThermostat(13, temp * 2);
  }

  public watch(): void {
    if (this.writeQueues.length) {
      setTimeout(this.watch.bind(this), 300);
      return;
    }
    this.update().finally(() => {
      setTimeout(this.watch.bind(this), 300);
    });
  }

  private writeToThermostat(register: number, value: number) {
    const retryAction = new RetriableAction(() => this.modbusMaster.writeSingleRegister(this.modbusAddr, register, value, 1), 50);
    this.writeQueues.add(register, retryAction);
  }

  private isHwDataChanged(data: number[]) {
    return (
      this.currentData[ThermostatRegister.ROOM_TEMP] !== data[ThermostatRegister.ROOM_TEMP] ||
      this.currentData[ThermostatRegister.ENABLED] !== data[ThermostatRegister.ENABLED] ||
      this.currentData[ThermostatRegister.TEMP_SETPOINT] !== data[ThermostatRegister.TEMP_SETPOINT]
    );
  }
}
