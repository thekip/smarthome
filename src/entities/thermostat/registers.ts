export enum ThermostatRegister {
  ENABLED = 0,
  FAN_SPEED = 1,
  MODE = 2,
  ROOM_TEMP = 3,
  TEMP_SETPOINT = 4,
  MANUAL_WEEKLY_PROG = 5,

  MINUTE_SECOND = 6,
  WEEK_HOUR = 7,
}
