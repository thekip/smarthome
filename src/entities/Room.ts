import { pick } from 'lodash';
import { EventEmitter } from '../libs/event-emitter/event-emitter';
import { Log } from '../libs/log';
import { AcUnit, AcMode as AC_MODES } from './AcUnit';
import chalk from 'chalk';
import { Dumper } from './Dumper';
import { Thermostat } from './thermostat/Thermostat';

export enum RoomChangeEmitter {
  thermostat = 'thermostat',
  api = 'api',
}

export interface RoomChangeData {
  emitter: RoomChangeEmitter;
}

let id = 0;

export interface RoomDto {
  tempSetpoint: number;
  enabled: boolean;
  ambientTemp: number;
  sync: boolean;
  id: number;
  dumperOpened: boolean;
}

export class Room {
  public onChange = new EventEmitter<RoomChangeData>();
  public id = id++;
  public sync: boolean = false;

  constructor(
    private thermostat: Thermostat,
    public dumper: Dumper,
    private ac: AcUnit,
  ) {
    thermostat.onChange.subscribe(() => {
      this.onThermostatChange();
    });
  }

  public updateDumperPosition() {
    if (!this.enabled) {
      return this.dumper.close();
    }

    let position = 'close';

    if (this.ac.mode === AC_MODES.COOL) {
      position = (this.ambientTemp - this.tempSetpoint) <= 0 ? 'close' : 'open';
    }

    if (this.ac.mode === AC_MODES.HEAT) {
      position = (this.tempSetpoint - this.ambientTemp) <= 0 ? 'close' : 'open';
    }

    if (position === 'close') {
      this.dumper.close();
    } else {
      this.dumper.open();
    }
  }

  public get isActive() {
    return this.enabled && this.dumper.isOpened;
  }

  public get enabled() {
    return this.thermostat.enabled;
  }

  public set enabled(value) {
    if (this.enabled === value) {
      return;
    }

    this.thermostat.setEnable(value);

    this.onChange.emit({
      emitter: RoomChangeEmitter.api,
    });
  }

  public get ambientTemp() {
    return this.thermostat.roomTemp;
  }

  public get tempSetpoint() {
    return this.thermostat.tempSetpoint;
  }

  public set tempSetpoint(temp) {
    if (this.tempSetpoint === temp) {
      return;
    }

    this.thermostat.setTempSetpoint(temp);

    this.onChange.emit({
      emitter: RoomChangeEmitter.api,
    });
  }

  public getDto(): RoomDto {
    const dto: RoomDto = pick(this, ['tempSetpoint', 'enabled', 'ambientTemp', 'sync', 'id']) as any;
    dto.dumperOpened = this.dumper.isOpened;
    return dto;
  }

  public setDto(dto: Partial<RoomDto>) {
    Object.assign(this, dto);
  }

  private onThermostatChange() {
    Log.info(chalk.cyan(`Room: thermostat is updated! ${this.id}: ` + this.thermostat.toString()));

    this.onChange.emit({
      emitter: RoomChangeEmitter.thermostat,
    });
  }
}
