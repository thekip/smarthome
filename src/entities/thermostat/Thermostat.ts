import { EventEmitter } from '../../libs/event-emitter';
import { ModbusAdapter, ThermostatModbusDto } from './modbus-adapter';
import { ThermostatRegister } from './registers';
import Bluebird = require('bluebird');

export const DEFAULT_SETPOINT = 24;

export class Thermostat {
  public enabled = false;
  public roomTemp: number = -1;
  public tempSetpoint = DEFAULT_SETPOINT;

  /**
   * Emits only if hardware data changed
   */
  public onChange = new EventEmitter();

  constructor(
    private adapter: ModbusAdapter,
  ) {
    adapter.onChange.subscribe((data) => this.handleThermostatChange(data));
  }

  private handleThermostatChange(data: ThermostatModbusDto) {
    this.enabled = data.enabled;
    this.tempSetpoint = data.tempSetpoint;
    this.roomTemp = data.roomTemp;

    this.onChange.emit();
  }

  public toString(): string {
    return `Enabled: ${this.enabled ? 'On' : 'Off'}; Setpoint: ${this.tempSetpoint} C; Room temp: ${this.roomTemp} C;`;
  }

  public setEnable(value: boolean): Bluebird<void> {
    this.enabled = !!value;
    return this.adapter.writeToThermostat(ThermostatRegister.ENABLED, this.enabled);

  }

  public setTempSetpoint(temp: number): Bluebird<void> {
    this.tempSetpoint = temp;
    return this.adapter.writeToThermostat(ThermostatRegister.TEMP_SETPOINT, temp);
  }
}
