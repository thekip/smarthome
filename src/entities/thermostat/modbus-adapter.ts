import { ModbusMaster } from 'modbus-rtu';
import { ModbusCrcError, ModbusResponseTimeout } from 'modbus-rtu/lib/errors';
import { DeviceConnection } from '../../libs/device-connection';
import { EventEmitter } from '../../libs/event-emitter';
import { Log } from '../../libs/log';
import { RetriableAction } from '../../libs/retriable-action';
import { QueueMap, WriteQueue } from '../../libs/write-queue';
import { ThermostatRegister } from './registers';
import { ValueMapper } from './value-mapper';
import Bluebird = require('bluebird');

export const ENABLED_VALUE = 165;
export const DISABLED_VALUE = 90;

export interface ThermostatModbusDto {
  roomTemp: number;
  enabled: boolean;
  tempSetpoint: number;
}

export class ModbusAdapter {
  private currentData: number[] = [];
  private writeQueues = new QueueMap(() => new WriteQueue());
  private connection = new DeviceConnection();

  /**
   * Emits only if hardware data changed
   */
  public onChange = new EventEmitter<ThermostatModbusDto>();

  constructor(
    private modbusMaster: ModbusMaster,
    private modbusAddr: number,
    noWatch = false,
  ) {
    this.connection.onConnectionLost.subscribe((err) => {
      Log.error('Lost connection to thermostat ' + this.modbusAddr, 'Error:', err.message);
    });

    this.connection.onConnectionRestore.subscribe(() => {
      Log.error('Connection restored. Thermostat ' + this.modbusAddr);
    });

    if (!noWatch) {
      this.watch();
    }
  }

  public update() {
    return this.modbusMaster.readHoldingRegisters(this.modbusAddr, 0, 6).then((data) => {
      this.connection.success();

      if (this.writeQueues.length) {
        // discard values from thermostat, if there are tasks
        return this.currentData;
      }

      // check, whether data is changed or not
      if (this.isHwDataChanged(data)) {
        this.onChange.emit({
          enabled: ValueMapper.fromModbus<boolean>(ThermostatRegister.ENABLED, data[ThermostatRegister.ENABLED]),
          roomTemp: ValueMapper.fromModbus<number>(ThermostatRegister.ROOM_TEMP, data[ThermostatRegister.ROOM_TEMP]),
          tempSetpoint: ValueMapper.fromModbus<number>(ThermostatRegister.TEMP_SETPOINT, data[ThermostatRegister.TEMP_SETPOINT]),
        });
      }

      this.currentData = data.slice(0); // clone data array
    }).catch(ModbusCrcError, ModbusResponseTimeout, (err: Error) => {
      this.connection.error(err);
    });
  }

  public watch(): void {
    if (this.writeQueues.length) {
      setTimeout(this.watch.bind(this), 300);
      return;
    }
    this.update().finally(() => {
      setTimeout(this.watch.bind(this), 300);
    });
  }

  public writeToThermostat(register: number, value: any): Bluebird<void> {
    const modbusValue = ValueMapper.toModbus(register, value);
    const retryAction = new RetriableAction(() => this.modbusMaster.writeSingleRegister(this.modbusAddr, register + 9, modbusValue, 1), 50);
    return this.writeQueues.add(register, retryAction);
  }

  private isHwDataChanged(data: number[]) {
    return (
      this.currentData[ThermostatRegister.ROOM_TEMP] !== data[ThermostatRegister.ROOM_TEMP] ||
      this.currentData[ThermostatRegister.ENABLED] !== data[ThermostatRegister.ENABLED] ||
      this.currentData[ThermostatRegister.TEMP_SETPOINT] !== data[ThermostatRegister.TEMP_SETPOINT]
    );
  }
}
