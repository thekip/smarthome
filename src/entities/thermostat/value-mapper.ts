import { DISABLED_VALUE, ENABLED_VALUE } from './modbus-adapter';
import { ThermostatRegister } from './registers';

export class ValueMapper {
  private static mappingsFrom: {[register: number]: (value: number) => any} = {
    [ThermostatRegister.ENABLED]: (value) => value === ENABLED_VALUE,
    [ThermostatRegister.TEMP_SETPOINT]: (value) => value / 2,
    [ThermostatRegister.ROOM_TEMP]: (value) => value / 2,
  };
  private static mappingsTo: {[register: number]: (value: number) => number} = {
    [ThermostatRegister.ENABLED]: (value) => !value ? DISABLED_VALUE : ENABLED_VALUE,
    [ThermostatRegister.TEMP_SETPOINT]: (value) => value * 2,
    [ThermostatRegister.ROOM_TEMP]: (value) => value * 2,
  };

  public static toModbus(register: number, value: any): number {
    return this.mappingsTo[register](value);
  }

  public static fromModbus<T>(register: number, value: number): T {
    return this.mappingsFrom[register](value);
  }
}
