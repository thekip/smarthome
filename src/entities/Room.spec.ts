import { EventEmitter } from '../libs/event-emitter';
import { AcMode, AcUnit } from './AcUnit';
import { Dumper } from './Dumper';
import { Room, RoomChangeEmitter } from './Room';
import { Thermostat } from './thermostat/Thermostat';

describe('Room', () => {
  function getDumperMock() {
    let isOpened = false;
    const dumper: Partial<Dumper> = {
      get isOpened() {
        return isOpened;
      },
      close: () => {
        isOpened = false;
      },
      open: () => {
        isOpened = true;
      },
    };

    return dumper as Dumper;
  }

  function getThermostatMock() {
    const thermostat: Partial<Thermostat> = {
      enabled: false,
      roomTemp: 28,
      tempSetpoint: 25,
      setEnable: jest.fn((value) => {
        thermostat.enabled = value;
      }),
      setTempSetpoint: jest.fn((value) => {
        thermostat.tempSetpoint = value;
      }),
      onChange: new EventEmitter(),
      toString: () => {
        return 'TESTING';
      },
    };

    return thermostat as Thermostat;
  }

  function testSetter(property: keyof Room, value: any, thermostatMethod: keyof Thermostat) {
    const thermostat = getThermostatMock();
    const dumper = getDumperMock();
    const room = new Room(thermostat as Thermostat, dumper as Dumper, {} as any);

    const handler = jest.fn();
    room.onChange.subscribe(handler);

    (room as any)[property] = value;
    expect(room[property]).toBe(value);

    expect(handler).toHaveBeenCalledWith({ emitter: RoomChangeEmitter.api });
    expect((thermostat[thermostatMethod] as any)).toHaveBeenCalledTimes(1);
  }

  it('When setpoint changed, should emit event and update thermostat ', () => {
    testSetter('tempSetpoint', 22, 'setTempSetpoint');
  });

  it('When `enabled` changed, should emit event and update thermostat ', () => {
    testSetter('enabled', true, 'setEnable');
  });

  test('Should emit event when thermostat changed', () => {
    const thermostat = getThermostatMock();
    const dumper = getDumperMock();
    const room = new Room(thermostat, dumper, {} as any);

    const handler = jest.fn();
    room.onChange.subscribe(handler);

    thermostat.onChange.emit();

    expect(handler).toHaveBeenCalledWith({ emitter: RoomChangeEmitter.thermostat });
  });

  test('Should be active only when thermostat enabled and dumper is open', () => {
    const thermostat = getThermostatMock();
    thermostat.enabled = true;
    const dumper = getDumperMock();
    const ac: Partial<AcUnit> = {
      mode: AcMode.COOL,
    };

    const room = new Room(thermostat, dumper, ac as AcUnit);
    room.updateDumperPosition();

    expect(room.isActive).toBeTruthy();

    thermostat.enabled = false;
    room.updateDumperPosition();
    expect(room.isActive).toBeFalsy();
  });

  interface DumperCase {
    enabled: boolean;
    roomTemp: number;
    tempSetpoint: number;
    expected: boolean;
    msg: string;
  }

  function testDumperControl(ac: AcUnit, cases: DumperCase[]) {
    const thermostat = getThermostatMock();
    const dumper = getDumperMock();

    const room = new Room(thermostat, dumper, ac);

    for (const spec of cases) {
      it(spec.msg, () => {
        thermostat.enabled = spec.enabled;
        thermostat.roomTemp = spec.roomTemp;
        thermostat.tempSetpoint = spec.tempSetpoint;

        room.updateDumperPosition();

        expect(dumper.isOpened).toBe(spec.expected);
      });
    }
  }

  describe('Dumper controlling in heating mode', () => {
    const ac: Partial<AcUnit> = {
      mode: AcMode.HEAT,
    };

    const cases: DumperCase[] = [
      {
        enabled: true,
        roomTemp: 22,
        tempSetpoint: 25,
        expected: true,
        msg: 'Setpoint higher then ambient temp, room enabled, dumper should be opened',
      },
      {
        enabled: false,
        roomTemp: 22,
        tempSetpoint: 25,
        expected: false,
        msg: 'Room disabled, dumper should be closed',
      },
      {
        enabled: true,
        roomTemp: 25,
        tempSetpoint: 25,
        expected: false,
        msg: 'Setpoint and ambient temp are equals, room enabled, dumper should be closed',
      },
      {
        enabled: true,
        roomTemp: 26,
        tempSetpoint: 25,
        expected: false,
        msg: 'Setpoint lower then ambient temp, room enabled, dumper should be closed',
      },
    ];

    testDumperControl(ac as AcUnit, cases);
  });

  describe('Dumper controlling in cooling mode', () => {
    const ac: Partial<AcUnit> = {
      mode: AcMode.COOL,
    };

    const cases: DumperCase[] = [
      {
        enabled: true,
        roomTemp: 25,
        tempSetpoint: 22,
        expected: true,
        msg: 'Setpoint lower then ambient temp, room enabled, dumper should be opened',
      },
      {
        enabled: false,
        roomTemp: 25,
        tempSetpoint: 22,
        expected: false,
        msg: 'Room disabled, dumper should be closed',
      },
      {
        enabled: true,
        roomTemp: 25,
        tempSetpoint: 25,
        expected: false,
        msg: 'Setpoint and ambient temp are equals, room enabled, dumper should be closed',
      },
      {
        enabled: true,
        roomTemp: 24,
        tempSetpoint: 25,
        expected: false,
        msg: 'Setpoint higher then ambient temp, room enabled, dumper should be closed',
      },
    ];

    testDumperControl(ac as AcUnit, cases);
  });

  it('should return DTO', () => {
    const thermostat = getThermostatMock();
    const dumper = getDumperMock();
    const room = new Room(thermostat, dumper, {} as any);

    expect(room.getDto()).toMatchSnapshot();
  });
});
