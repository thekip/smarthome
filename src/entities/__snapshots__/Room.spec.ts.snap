// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room should return DTO 1`] = `
Object {
  "ambientTemp": 28,
  "dumperOpened": false,
  "enabled": false,
  "id": 6,
  "sync": false,
  "tempSetpoint": 25,
}
`;
